src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -395,15 +400,20 @@ public ThreadPoolWorkQueue()
             ThreadPoolWorkQueueThreadLocals.threadLocals ??
             (ThreadPoolWorkQueueThreadLocals.threadLocals = new ThreadPoolWorkQueueThreadLocals(this));
 
+        internal bool ThreadRequestNeeded(int count) => (count < ThreadPoolGlobals.processorCount) &&
+            ((count < workItems.Count) || (WorkStealingQueueList.wsqActive > 0));
```

Comment:
@kouvel 

Added `bool ThreadRequestNeeded(int count)`

The `(count < ThreadPoolGlobals.processorCount)` is the baseline.
The `(count < workItems.Count)` makes this perform better than #13408 
The `(WorkStealingQueueList.wsqActive > 0)` disable the optimization if WSQ is in use.  The WSQ cannot be calculated efficiently.  
My preference would be to use `(count < workItems.Count + WorkStealingQueueList.wsqActive)` instead of 
`((count < workItems.Count) || (WorkStealingQueueList.wsqActive > 0))`.  However since I am not testing WSQ, I have no justification only a theoretical argument.

Using `(count < Math.Max(ThreadPoolGlobals.processorCount - runningCount, 1))` would make this patch equivalent with #13386.  It would require more overhead to track running thread.  Essentially an extra Interlock.Decrement() on thread exit.  I am tempted to defer this extra overhead and complexity until the PR to replace UnfairSemaphore is written/tested.